Added changelist docs for alpha 4 (#325)
gchq/gaffer-doc: docs/federation-changes.md

    "mergeFunction": {
        "class": "uk.gov.gchq.gaffer.federatedstore.util.ConcatenateMergeFunction"
    }
}
```

### Required parameter: operation

This is the Operation you wish to be federated to the subgraphs. This can be a single Operation or an OperationChain. If you use an OperationChain, then the whole chain will be sent together to the subgraphs.  

### Optional parameter: graphIds

This is a list of graph IDs which you want to send the operation to. If you have an OperationChain and wish to send each operation to a different graph, then each operation in the chain will need to be wrapped in a FederatedOperation with the `graphIds` set to the required graph. If instead you want to send the whole OperationChain to the same graphs, then this field can be set on one FederatedOperation where the `operation` is an OperationChain.  

If this parameter is not provided, then the default graph IDs for the store will be used. This is the `storeConfiguredGraphIds` for the store if set, if not set then all graphIds will be used.

### Optional parameter: mergeFunction

The `mergeFunction` parameter is the BiFunction you want to use when merging the results from the subgraphs.  

> If this parameter is not provided, then the store's `storeConfiguredMergeFunctions` will be used, which have default values but can be overriden. For example, by default when GetElements is used as the operation inside a FederatedOperation, then `ApplyViewToElementsFunction` will be used to merge the results. Instead, this can be overriden, like in the example to something like `ConcatenateMergeFunction`, or even a custom Function.

Reviewer: Didn't the default behaviour here change compared with using a `FederatedOperationChain` before? I thought concatenating had been the default and now it is not? If so, this should be documented in more detail, might be considered a breaking change.

Author: Yes I was going to doc that more in detail in the full FederatedStore docs but I will add some info on this change as it could be breaking.

Author: I added some basic docs on what has changed, roughly how results may differ, and how to get old behaviour. I think more detail on each merge function should be given in the core Federated Store docs refactor and then linked to from here.

Reviewer: This is great, exactly the right amount of detail. When the full documentation has been refactored we can add some links to that.

# Federated Store Changes

This page contains information on the changes to Gaffer's Federated Store. This functionality was introduced in version `2.0.0-alpha-0.4` of Gaffer.  
The main changes were the addition of the [Federated Operation](#the-federated-operation), and a change to how results are [merged](#default-results-merging) by default.

## The Federated Operation

The `FederatedOperationChain` was removed and replaced with a new Operation, the `FederatedOperation`. This was added to improve the control you have over how operations are federated.  
The Federated Operation has 3 key parameters: `operation`, `graphIds` and `mergeFunction`:
``` json
{
    "class": "uk.gov.gchq.gaffer.federatedstore.operation.FederatedOperation",
    "operation": {
        "class": "uk.gov.gchq.gaffer.operation.impl.get.GetAllElements"
    },
    "graphIds": [ "graphA", "graphB" ],
    "mergeFunction": {
        "class": "uk.gov.gchq.gaffer.federatedstore.util.ConcatenateMergeFunction"
    }
}
```

### Required parameter: operation

> This is the Operation you wish to be federated to the subgraphs. This can be a single Operation or an OperationChain. If you use an OperationChain, then the whole chain will be sent together to the subgraphs.  

Reviewer 2: ```suggestion
This is the Operation you wish to be federated to the subgraphs. This can be a single Operation or an OperationChain. If you use an OperationChain, then all the operations in the chain will be sent together to the subgraphs.  
```

### Optional parameter: graphIds

> This is a list of graph IDs which you want to send the operation to. If you have an OperationChain and wish to send each operation to a different graph, then each operation in the chain will need to be wrapped in a FederatedOperation with the `graphIds` set to the required graph. If instead you want to send the whole OperationChain to the same graphs, then this field can be set on one FederatedOperation where the `operation` is an OperationChain.  

Reviewer 2: I think it's a little muddy here. The regular usage of FedOp with a opChain was already established in the previous chapter. I'd only mention new alternative part here. Also it doesn't need to be per operation. A section could be multiple operation going to the same graph.

```suggestion
This is a list of graph IDs which you want to send the operation to. If you have an OperationChain and require sending different sections of the chain to different graphs, then you require a chain of FederatedOperations, with one FederatedOperation for each section of the chain sharing the same `graphIds`.
```

> If this parameter is not provided, then the default graph IDs for the store will be used. This is the `storeConfiguredGraphIds` for the store if set, if not set then all graphIds will be used.

Reviewer 2: ```suggestion
If the user does not specify graphIds then the `storeConfiguredGraphIds` for that store will be used. If the admin does not specify the 'storeConfiguredGraphIds' then all graphIds will be used.
```

### Optional parameter: mergeFunction

The `mergeFunction` parameter is the BiFunction you want to use when merging the results from the subgraphs.  

> If this parameter is not provided, then the store's `storeConfiguredMergeFunctions` will be used, which have default values but can be overriden. For example, by default when GetElements is used as the operation inside a FederatedOperation, then `ApplyViewToElementsFunction` will be used to merge the results. Instead, this can be overriden, like in the example to something like `ConcatenateMergeFunction`, or even a custom Function.

Reviewer 2: ```suggestion
If the user does not specify a `mergeFunction` then a `mergeFunction` for the operation will be selected from `storeConfiguredMergeFunctions` for that store. If the admin has not specified the `storeConfiguredMergeFunctions` it will contain pre-populated `mergeFunctions`. Lastly if a suitable `mergeFunction` is not found then a default `ConcatenateMergeFunction` is used. In example, when GetElements is used as the operation inside a FederatedOperation and the user hasn't specified a `mergeFunction` the pre-populated `ApplyViewToElementsFunction` will be selected from `storeConfiguredMergeFunctions`. Unless the admin configured `storeConfiguredMergeFunctions` to use something else.
```

            }
        },
        {
            "class": "ExampleOperation2",
            "options": {
                "gaffer.federatedstore.operation.graphIds": "graphB"
            }
        }
    ]
}
```

#### New FederatedOperation graphIds inside an OperationChain
```json
{
    "class": "uk.gov.gchq.gaffer.operation.OperationChain",
    "operations": [
        {
            "class": "uk.gov.gchq.gaffer.federatedstore.operation.FederatedOperation",
            "operation": {
                "class": "ExampleOperation1"
            },
            "graphIds": [ "graphA" ]
        },
        {
            "class": "uk.gov.gchq.gaffer.federatedstore.operation.FederatedOperation",
            "operation": {
                "class": "ExampleOperation2"
            },
            "graphIds": [ "graphB" ]
        }
    ]
}
```

### Breaking change

#### No longer supported graphIds option on an OperationChain
Previously, if you wanted to send an entire OperationChain to a specific graph, you could use the `graphIds` option on the chain, like so:  
```json
{
    "class": "uk.gov.gchq.gaffer.operation.OperationChain",
    "operations": [
        {

>             "class": "uk.gov.gchq.gaffer.operation.impl.get.GetAllElements"

Reviewer 2: can you do a chain with 2 operations at least please.

        }
    ],
    "options": {
        "gaffer.federatedstore.operation.graphIds": "graphA"
    }
}
```

> #### New FederatedOperation graphIds on an OperationChain

Reviewer 3: I'm not sure the this makes sense, but i could be misunderstanding.

> #### New FederatedOperation graphIds on an OperationChain
> This is no longer supported, and you should instead wrap the chain in a FederatedOperation

Makes it sound like 'New FederatedOperation graphIds on an OperationChain' isn't supported, but i'm assuming that it is supported and this is the new way of doing it.

{
    "class": "uk.gov.gchq.gaffer.federatedstore.operation.FederatedOperation",
    "operation": {
        "class": "uk.gov.gchq.gaffer.operation.OperationChain",
        "operations": [
            {
                "class": "uk.gov.gchq.gaffer.operation.impl.get.GetAllElements"
            }
        ]
    },
    "graphIds": [ "graphA" ]
}
```

## Default results merging

As described above, FederatedStores now have `storeConfiguredMergeFunctions` that dictate how the FederatedStore will merge results from different subgraphs dependent on the Operation.  

In places, these new defaults do differ from previous behaviour, hence results will too. This can be overriden on a per Operation basis using the `mergeFunction` parameter described above, or a per store basis by overriding `storeConfiguredMergeFunctions`.  
The previous behaviour was that all Operation results were concatenated together, this is now a mergeFunction within Gaffer called `ConcatenateMergeFunction`. Therefore, if you wanted a FederatedOperation to use this old behaviour, you can set the `mergeFunction` to `ConcatenateMergeFunction` (as shown [above](#the-federated-operation)).  

### New Merge function examples

> By default, `GetElements` results will be merged with `ApplyViewToElementsFunction`. This uses the View from the operation and applies it to all of the results, meaning the results are now aggregated and filtered using the Schema too. This makes the results look like they came from one graph, rather than getting back a list of Elements from different subgraphs.  

Reviewer 2: ```suggestion
By default, `GetElements` results will be merged with `ApplyViewToElementsFunction`. This uses the View from the operation and applies it to all of the results, meaning the results are now re-aggregated and re-filtered using the Schema, locally in the FederatedStore. This makes the results look like they came from one graph, rather than getting back a list of Elements from different subgraphs.
```

> By default, `GetTraits` results will be merged with `CollectionIntersect`. This returns the intersection of common store traits from the subgraphs, rather than all of the traits concatenate, meaning it is a better representation of traits all the subgraphs have.  

Reviewer 2: ```suggestion
By default, `GetTraits` results will be merged with `CollectionIntersect`. This still returns the intersection of common store traits from the subgraphs, but from a BiFunction rather than a Handler.  
```

It did this before, now its done in a function rather than a FederatedTraitHandler.

### Default storeConfiguredMergeFunctions

| Operation         | Merge function              |
|-------------------|-----------------------------|
| GetElements       | ApplyViewToElementsFunction |
| GetAllElements    | ApplyViewToElementsFunction |
| GetSchema         | MergeSchema                 |
| GetTraits         | CollectionIntersect         |
| others            | ConcatenateMergeFunction    |

## Other changes

> Previously, if the FederatedStore received an Operation it did not have an OperationHandler for locally, the Operation would fail to execute. Now, the Operation will just be handed over to the subgraphs to be executed, as they may have the correct OperationHandlers. This should make it easier to use a ProxyStore with a FederatedStore, as the FederatedStore will no longer need to have every OperationHandler present on the remote ProxyStore.  

Reviewer 2: ```suggestion
```

There was Large Umbrella Handlers which catch and forward the operation to subgraphs. This type of erroring never happened.

By default, `GetSchema` results will be merged with `MergeSchema`. This returns an aggregated schema from the subgraphs, unless there is a conflict. This behaviour is the same, but now it can be overriden. For example, you may wish to use the `ConcatenateMergeFunction` if there is a schema conflict.  

### Default storeConfiguredMergeFunctions

| Operation         | Merge function              |
|-------------------|-----------------------------|
| GetElements       | ApplyViewToElementsFunction |
| GetAllElements    | ApplyViewToElementsFunction |
| GetSchema         | MergeSchema                 |
| GetTraits         | CollectionIntersect         |
| others            | ConcatenateMergeFunction    |

## Removal of FederatedOperationChain

The FederatedOperationChain has been removed, and where you would have used it before you should instead use a FederatedOperation with an OperationChain inside.  

This is useful if you have an OperationChain and want to send different parts of the chain to different graphs.

#### Individually sending a sequence of Operations to a subgraph

> You could send a sequence of operations within one chain to the same subgraph using `graphIds`, however, this is not always effecient:

Reviewer 3: ```suggestion
You could send a sequence of operations within one chain to the same subgraph using `graphIds`, however, this is not always efficient:
```

```json
{
    "class": "uk.gov.gchq.gaffer.operation.OperationChain",
    "operations": [
        {
            "class": "uk.gov.gchq.gaffer.federatedstore.operation.FederatedOperation",
            "operation": {
                "class": "ExampleOperation1"
            },
            "graphIds": [ "graphA" ]
        },
        {
            "class": "uk.gov.gchq.gaffer.federatedstore.operation.FederatedOperation",
            "operation": {
                "class": "ExampleOperation2"
            },
            "graphIds": [ "graphA" ]
        },
        {
            "class": "uk.gov.gchq.gaffer.federatedstore.operation.FederatedOperation",
            "operation": {
                "class": "ExampleOperation3"
            },
            "graphIds": [ "graphB" ]
        }
    ]
}
```

#### Removed FederatedOperationChain sending a sequence of operations to a subgraph

> It is more effecient to group together sequences of Operations that will go to the same subgraph.

Reviewer 3: ```suggestion
It is more efficient to group together sequences of Operations that will go to the same subgraph.
```
